chainer/optimizers/adam.py

```diff
@@ -81,8 +92,15 @@ def update_core_cpu(self, param):
 
         m += (1 - hp.beta1) * (grad - m)
         v += (1 - hp.beta2) * (grad * grad - v)
-        param.data -= hp.eta * (self.lr * m / (np.sqrt(v) + hp.eps) +
-                                hp.weight_decay_rate * param.data)
+
+        if hp.amsgrad:
+            vhat = self.state['vhat']
+            vhat = np.maximum(vhat, v)
+            param.data -= hp.eta * (self.lr * m / (np.sqrt(vhat) + hp.eps) +
+                                    hp.weight_decay_rate * param.data)
+        else:
+            param.data -= hp.eta * (self.lr * m / (np.sqrt(v) + hp.eps) +
+                                    hp.weight_decay_rate * param.data)
```

Comment:
Do you think it's possible to reduce code duplication by maybe introducing another (local) variable?

Reply:
yes sure... I can do that let me push it in a few mins